### PR TITLE
[Imgtool] C++-ified a few more classes

### PR DESCRIPTION
**Pull request by npwoods**
imgtool_partition ==> imgtool::partition
imgtool_directory ==> imgtool::directory

### CODE REVIEW

File: src/tools/imgtool/imgtool.h [hunk 1]
```diff
@@ -133,56 +133,132 @@ namespace imgtool
...
+		imgtoolerr_t(*m_list_forks)   (imgtool::partition *partition, const char *path, imgtool_forkent *ents, size_t len);
+		imgtoolerr_t(*m_create_dir)   (imgtool::partition *partition, const char *path);
+		imgtoolerr_t(*m_delete_dir)   (imgtool::partition *partition, const char *path);
+		imgtoolerr_t(*m_list_attrs)   (imgtool::partition *partition, const char *path, UINT32 *attrs, size_t len);
+		imgtoolerr_t(*m_get_attrs)    (imgtool::partition *partition, const char *path, const UINT32 *attrs, imgtool_attribute *values);
+		imgtoolerr_t(*m_set_attrs)    (imgtool::partition *partition, const char *path, const UINT32 *attrs, const imgtool_attribute *values);
+		imgtoolerr_t(*m_attr_name)    (UINT32 attribute, const imgtool_attribute *attr, char *buffer, size_t buffer_len);
+		imgtoolerr_t(*m_get_iconinfo) (imgtool::partition *partition, const char *path, imgtool_iconinfo *iconinfo);
+		imgtoolerr_t(*m_suggest_transfer)(imgtool::partition *partition, const char *path, imgtool_transfer_suggestion *suggestions, size_t suggestions_length);
+		imgtoolerr_t(*m_get_chain)    (imgtool::partition *partition, const char *path, imgtool_chainent *chain, size_t chain_size);
```
Comment (cuavas) [inline src/tools/imgtool/imgtool.h]: These should be replaced with `std::function` to allow more flexibility.

**Comment by npwoods** [comment]
Is `std::function` really considered to be that great a thing?

Using Compiler Explorer (https://gcc.godbolt.org/), I was surprised at the amount of overhead `std::function` seemed to entail.  I would have thought that the program below would have optimized down to a mere function pointer invocation.

```
typedef void (*foo_ptr)(int a, char *b, bool c);
extern foo_ptr get_foo_ptr();

int main(int argc, char *argv[])
{
	std::function<void(int, char*, bool)> my_foo_ptr = get_foo_ptr();
	my_foo_ptr(argc, argv[0], argc == 3);
	return 0;
}
```

**Comment by cuavas** [comment]
Calling a `std::function` can be optimised to a single vtable dispatch, but assigning to one necessarily involves allocation because it needs to create a wrapper for whatever callable object you supply and store it somewhere. It's a great thing because it's very flexible - it lets you use a function pointer, `std::bind`, a lambda expression or any object with `operator()`. A function pointer lacks this flexibility. For situations where you don't need to assign to the variable too and you can live with the hit to locality, the flexibility of `std::function` is a clear win. When it's in the critical path, we have our delegate templates that have better locality and lower overhead at the cost of being far less flexible.

**Comment by cuavas** [comment]
*don't need to assign to the variable too often

I accidentally a word again

**Comment by galibert** [comment]
Hmmm vas, not sure if you looked the -O9 asm, but it's not really nice.
And for some reason the invocation is not inlined...

  OG.

**Comment by cuavas** [comment]
I know it isn't fully inlined - it can't be because the specialised wrapper for the function object is behind a pointer and has a vtable. It's the cost of flexibility. But using naked function pointers prevents use of stateful lambdas, `std::bind` and arbitrary objects with `operator()`. It's not like this code is in the memory system or something.

**Comment by npwoods** [comment]
Fair enough - I can submit a follow up patch to address this once #1500 is merged.

**Pull request merged**